bedrock/firefox/views.py

```diff
@@ -119,15 +117,33 @@ def stub_attribution_code(request):
         codes['medium'] = '(none)'
 
     codes['timestamp'] = str(int(time()))
-    code = '&'.join('='.join(attr) for attr in codes.items())
-    code = querystringsafe_base64.encode(code)
+    response = HttpResponseJSON(sign_attribution_codes(codes))
+    patch_response_headers(response, 300)  # 5 min
+    return response
+
+
+def sign_attribution_codes(codes):
+    """
+    Take the attribution codes and return the base64 encoded string
+    respecting max length and HMAC signature.
+    """
+    key = settings.STUB_ATTRIBUTION_HMAC_KEY
+    code = b64_encode_codes(codes)
+    while len(code) > settings.STUB_ATTRIBUTION_MAX_LEN:
+        # remove 5 char at a time
+        codes['campaign'] = codes['campaign'][:-5] + '_'
+        code = b64_encode_codes(codes)
```

Comment:
Could this potentially result in an infinite loop if `codes['campaign']` gets cut to zero characters, but the string remains longer than the max (due to some other param being long)?

Reply:
ooooooo... great catch.. yes. I'll update.